h/notification/reply_template.py

```diff
@@ -106,17 +98,24 @@ def generate_notifications(request, annotation, action):
     if parent_id is None:
         return
     parent = storage.fetch_annotation(request, parent_id)
-    if parent is None or 'user' not in parent:
+    if parent is None or not annotation.userid:
+        return
+
+    # Don't send reply notifications to the author of the parent annotation if
+    # the reply is private (the author of the parent annotation won't be able
+    # to read the reply).
+    if not annotation.shared:
         return
 
-    # We don't send replies to the author of the parent unless they're going to
-    # be able to read it. That means there must be some overlap between the set
-    # of effective principals of the parent's author, and the read permissions
-    # of the reply.
-    child_read_permissions = annotation.get('permissions', {}).get('read', [])
-    parent_principals = auth.effective_principals(parent['user'], request)
-    read_principals = translate_annotation_principals(child_read_permissions)
-    if not set(parent_principals).intersection(read_principals):
+    # Don't send reply notifications to the author of the parent annotation if
+    # the reply is in a group that the author of the parent annotation isn't
+    # (the author of the parent annotation won't be able to read the reply).
+    parent_principals = auth.effective_principals(parent.userid, request)
+    if annotation.groupid == '__world__':
```

Comment:
This seems a bit fragile. Might be better to fetch the current authorization policy and use `policy.permits(annotation, parent_principals, 'read')`. That would probably eliminate some code higher up here, too.